### PR TITLE
[WIP] ctc_tdnn baseline recipe for Hi-Mia wake word detection

### PR DESCRIPTION
**Pull request by glynpu**
AUC of Himia-Aishell test set
image

AUC of HiMia-CW testset.  
According to https://arxiv.org/pdf/2201.00167.pdf, 
Confusion words(CW) test set is a more difficult test set. So the AUC is lower.

image

### CODE REVIEW

**Comment by danpovey** [comment]
Cool!  Sorry I don't think I have time/energy to go through this in detail, does anyone want to take responsibility for checking over it (to the extent that they feel able), and merging it?

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+            default=True,
+            help="When enabled, select noise from MUSAN and mix it"
+            "with training dataset. ",
+        )
+
+        group.add_argument(
+            "--input-strategy",
+            type=str,
+            default="PrecomputedFeatures",
+            help="PrecomputedFeatures",
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/asr_datamodule.py:193]: Could you document what possible values it can be?

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+        )
+        group.add_argument(
+            "--train-channel",
+            type=str,
+            default="_7_01",
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/asr_datamodule.py:198]: Could you document what possible values it can be?

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+                    duration_factor=self.args.duration_factor, gap=self.args.gap
+                )
+            ] + transforms
+
+        input_transforms = []
+        if self.args.enable_spec_aug:
+            logging.info("Enable SpecAugment")
+            logging.info(f"Time warp factor: {self.args.spec_aug_time_warp_factor}")
+            # Set the value of num_frame_masks according to Lhotse's version.
+            # In different Lhotse's versions, the default of num_frame_masks is
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/asr_datamodule.py:254]: I suggest that we remove such checks. Any recent versions of lhotse should work.

The checks are added more than a year ago when the changes made by mingshuang were not merged.

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py (hunk 2 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/asr_datamodule.py:197]: Thanks. fixed.

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py (hunk 3 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/asr_datamodule.py]: Thanks. fixed

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py (hunk 1 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/asr_datamodule.py]: Thanks. fixed.

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+            potential values. Currently, Only "_7_01" is verified.
+            """,
+        )
+        group.add_argument(
+            "--dev-channel",
+            type=str,
+            default="_7_01",
+            help="""channel of HI_MIA dev dataset.
+            All channels are used if it is set "all".
+            Please refer to state 6 in prepare.sh for its meaning and other
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/asr_datamodule.py:210]: state -> stage

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+        logging.info("About to get train cuts")
+        train_cuts_file = (
+            f"cuts_train_himia{self.args.train_channel}-aishell-shuf.jsonl.gz"
+        )
+        if "all" == self.args.train_channel:
+            train_cuts_file = "cuts_train_himia-aishell-shuf.jsonl.gz"
+        return load_manifest_lazy(self.args.manifest_dir / f"{train_cuts_file}")
+
+    @lru_cache()
+    def aishell_test_cuts(self) -> CutSet:
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/asr_datamodule.py:402]: Shall we delete it?

File: egs/himia/wuw/README.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# Pretrained models and related logs/results.
+
+## ctc tdnn baseline
+
+AUC results for different epochs could be found at <https://huggingface.co/GuoLiyong/himia_ctc_tdnn_baseline/tree/main>
+
+E.g. for epoch 15 and avg 1, result log file is: <https://huggingface.co/GuoLiyong/himia_ctc_tdnn_baseline/blob/main/exp_max_duration_100/post/epoch_15-avg_1/log/log-auc-himia_aishell-2023-03-16-17-42-14>
+
+Corresponding ROC curve is: <https://huggingface.co/GuoLiyong/himia_ctc_tdnn_baseline/blob/main/exp_max_duration_100/post/epoch_15-avg_1/himia_aishell.pdf>
```
Comment (csukuangfj) [inline egs/himia/wuw/README.md:9]: Could you create a RESULTS.md and put some images of ROC curves in it for ease of reference.

File: egs/himia/wuw/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
```
Comment (csukuangfj) [inline egs/himia/wuw/README.md:10]: Could you also document the number of model parameters in RESULTS.md?

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+class WakeupWordTokenizer(object):
+    def __init__(
+        self,
+        wakeup_word: str = "",
+        wakeup_word_tokens: List[int] = None,
+    ) -> None:
+        """
+        Args:
+          wakeup_word: content of positive samples.
+            A sample will be treated as a negative sample unless its context
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:32]: context-> content ?

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+import torch
+
+from typing import List, Tuple
+
+
+class WakeupWordTokenizer(object):
+    def __init__(
+        self,
+        wakeup_word: str = "",
+        wakeup_word_tokens: List[int] = None,
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:27]: List[int] -> Optional[List[int]]

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    ) -> None:
+        """
+        Args:
+          wakeup_word: content of positive samples.
+            A sample will be treated as a negative sample unless its context
+            is exactly the same to key_words.
+          wakeup_word_tokens: A list if int represents token ids of wakeup_word.
+            For example: the pronunciation of "你好米雅" is
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:35]: if -> of

represents -> representing

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+              m   6
+              y   7
+              a   8
+            Then wakeup_word_tokens for "你好米雅" is:
+             n  i  h  ao m  i  y  a
+            [2, 3, 4, 5, 6, 3, 7, 8]
+        """
+        super().__init__()
+        assert wakeup_word is not None
+        assert wakeup_word_tokens is not None
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:53]: if it cannot be None, please don't use None as its default value.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+          The last one is number_positive_samples,
+          used to track proportion of positive samples in each batch.
+        """
+        batch_token_ids = []
+        target_lengths = []
+        number_positive_samples = 0
+        for utt_text in texts:
+            if utt_text == self.wakeup_word:
+                batch_token_ids.append(self.wakeup_word_tokens)
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:92]: I suggest that you replace append with extend and remove itertools below.

It would simplify the code a lot.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        """Convert a list of texts to a list of k2.Fsa based texts.
+
+        Args:
+          texts:
+            It is a list of strings,
+            each element is a reference text for an audio.
+        Returns:
+          Return a tuple of 3 elements.
+          The first one is torch.Tensor(List[List[int]]),
+          each List[int] is tokens sequence for each a reference text.
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:79]: each a -> each

File: egs/himia/wuw/ctc_tdnn/tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        self.wakeup_word = wakeup_word
+        self.wakeup_word_tokens = wakeup_word_tokens
+        self.positive_number_tokens = len(wakeup_word_tokens)
+        self.negative_word_tokens = [1]
+        self.negative_number_tokens = 1
+
+    def texts_to_token_ids(
+        self, texts: List[str]
+    ) -> Tuple[torch.Tensor, torch.Tensor, int]:
+        """Convert a list of texts to a list of k2.Fsa based texts.
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/tokenizer.py:70]: The return value is not related to k2.Fsa, I think.

File: egs/himia/wuw/ctc_tdnn/graph.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def ctc_trivial_decoding_graph(wakeup_word_tokens: List[int]):
+    """
+    A graph starts with blank/unknown and following by wakeup word.
+
+    Args:
+      wakeup_word_tokens: A sequence of token ids corresponding wakeup_word.
+      It should not contain 0 and 1.
+      We assume 0 is for blank and 1 is for unknown.
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/graph.py:29]: Could you also document the return value?

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py (hunk 5 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/asr_datamodule.py:402]: No, aishell test is used as negative samples test set.

File: egs/himia/wuw/ctc_tdnn/asr_datamodule.py (hunk 4 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/asr_datamodule.py]: Thanks. fixed

File: egs/himia/wuw/README.md (hunk 6 above)
Comment (glynpu) [reply egs/himia/wuw/README.md]: Good suggestions. Added. Thanks.

File: egs/himia/wuw/README.md (hunk 7 above)
Comment (glynpu) [reply egs/himia/wuw/README.md:10]: added. Thanks.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py (hunk 8 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/tokenizer.py]: fixed. Thanks.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py (hunk 11 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/tokenizer.py:53]: fixed. thanks.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py (hunk 12 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/tokenizer.py]: fixed. Thanks.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py (hunk 13 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/tokenizer.py]: fixed. Thanks.

File: egs/himia/wuw/ctc_tdnn/graph.py (hunk 15 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/graph.py:29]: fixed. Thanks.

File: egs/himia/wuw/ctc_tdnn/tokenizer.py (hunk 10 above)
Comment (glynpu) [reply egs/himia/wuw/ctc_tdnn/tokenizer.py:35]: fixed. thanks.

File: egs/himia/wuw/ctc_tdnn/graph.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    assert 0 not in wakeup_word_tokens
+    assert 1 not in wakeup_word_tokens
+    assert len(wakeup_word_tokens) >= 2
+    keyword_ilabel_start = wakeup_word_tokens[0]
+    fst_graph = ""
+    for non_wake_word_token in range(keyword_ilabel_start):
+        fst_graph += f"0 0 {non_wake_word_token} 0\n"
+    cur_state = 1
+    for token_idx in range(len(wakeup_word_tokens) - 1):
+        token = wakeup_word_tokens[token_idx]
```
Comment (csukuangfj) [inline egs/himia/wuw/ctc_tdnn/graph.py:44]:
```suggestion
    for token in wakeup_word_tokens[:-1]:
```
is much simpler.
